core/shared/src/main/scala/scalaz/zio/ZManaged.scala

```diff
@@ -73,93 +213,516 @@ final case class ZManaged[-R, +E, +A](reserve: ZIO[R, E, Reservation[R, E, A]])
+   * inner effect, yielding the value of the inner effect.
+   *
+   * This method can be used to "flatten" nested effects.
+    **/
+  final def flatten[R1 <: R, E1 >: E, B](implicit ev: A <:< ZManaged[R1, E1, B]): ZManaged[R1, E1, B] =
+    self.flatMap(ev)
 
   /**
-   * Named alias for `<*`.
-   */
-  final def zipLeft[R1 <: R, E1 >: E, A1](that: ZManaged[R1, E1, A1]): ZManaged[R1, E1, A] =
-    self <* that
+   * Flip the environment and result
+    **/
+  final def flip: ZManaged[R, A, E] =
+    self.foldM(ZManaged.succeed, ZManaged.fail)
 
-  final def zipWith[R1 <: R, E1 >: E, A1, A2](that: ZManaged[R1, E1, A1])(f: (A, A1) => A2): ZManaged[R1, E1, A2] =
-    flatMap(r => that.map(r1 => f(r, r1)))
+  /**
+   * Flip the environment and result and then apply an effectful function to the effect
+    **/
+  final def flipWith[R1, A1, E1](f: ZManaged[R, A, E] => ZManaged[R1, A1, E1]): ZManaged[R1, E1, A1] =
+    f(self.flip).flip
 
-  final def <*>[R1 <: R, E1 >: E, A1](that: ZManaged[R1, E1, A1]): ZManaged[R1, E1, (A, A1)] =
-    zipWith(that)((_, _))
+  /**
+   * Folds over the failure value or the success value to yield an effect that
+   * does not fail, but succeeds with the value returned by the left or right
+   * function passed to `fold`.
+   */
+  final def fold[B](failure: E => B, success: A => B): ZManaged[R, Nothing, B] =
+    foldM(failure.andThen(ZManaged.succeed), success.andThen(ZManaged.succeed))
 
   /**
-   * Named alias for `<*>`.
+   * A more powerful version of `foldM` that allows recovering from any kind of failure except interruptions.
    */
-  final def zip[R1 <: R, E1 >: E, A1](that: ZManaged[R1, E1, A1]): ZManaged[R1, E1, (A, A1)] =
-    self <*> that
+  final def foldCauseM[R1 <: R, E1, A1](
+    failure: Cause[E] => ZManaged[R1, E1, A1],
+    success: A => ZManaged[R1, E1, A1]
+  ): ZManaged[R1, E1, A1] =
+    sandbox.foldM(failure, success)
 
-  final def zipWithPar[R1 <: R, E1 >: E, A1, A2](that: ZManaged[R1, E1, A1])(f0: (A, A1) => A2): ZManaged[R1, E1, A2] =
-    ZManaged[R1, E1, A2] {
+  /**
+   * Recovers from errors by accepting one effect to execute for the case of an
+   * error, and one effect to execute for the case of success.
+   */
+  final def foldM[R1 <: R, E2, B](
+    failure: E => ZManaged[R1, E2, B],
+    success: A => ZManaged[R1, E2, B]
+  ): ZManaged[R1, E2, B] =
+    ZManaged[R1, E2, B] {
       Ref.make[ZIO[R1, Nothing, Any]](IO.unit).map { finalizers =>
         Reservation(
           acquire = {
-            val left =
-              self.reserve.flatMap(res => finalizers.update(fs => res.release *> fs).const(res)).uninterruptible
-            val right =
-              that.reserve.flatMap(res => finalizers.update(fs => res.release *> fs).const(res)).uninterruptible
-
-            left.flatMap(_.acquire).zipWithPar(right.flatMap(_.acquire))(f0)
+            val direct =
+              ZIO.uninterruptibleMask { restore =>
+                restore(self.reserve)
```

Comment:
I'm totally with you when comes to the finalizers, but this should already be the case here if I understand uninterrutibleMask correctly.

I have one general question about the semantics of reserve. What kind of effects should be placed into it vs acquire?
For examble ZManaged.make pushes the entire allocation into reserve while ZManaged.fromEffect does the effect in acquire. What is the difference there?

Reply:
So the point of the reserve/acquire split is to allow for models of resource acquisition where you have a (short) reservation stage and a potentially long acquire stage. The acquire stage is interruptible in that it is possible to formulate a release action that works correctly whether or not `acquire` completed.

We haven't converted `Semaphore` to use it, but it works exactly this way: there's a preparation step (`prepare`), which is short, and an acquisition step (`acquire`), which is potentially infinite. `acquire` is safely interruptible because the resource release (`restore` in `prepare`) would do the right thing even if the acquisition was interrupted.

Now `ZManaged` supports both of these models. Note that the `use` function will bracket on the *reservation*, not on the acquisition; this means only the reservation would run uninterruptibly. So if your resource can split between reserve/acquire, you can use the `ZManaged.reserve` constructor or the main constructor. If you can't, you use `ZManaged.make`, which stuffs the entire acquisition into the reserve.

Regarding `ZManaged.fromEffect`, it's a good question if those effects should run uninterruptibly. I lean to run them interruptibly and noting about this in the scaladoc.

Reply:
Also, note that in `uninterruptibleMask`, only parts that are *not* under the `restore` function are guaranteed to run uninterruptibly. Anything under `restore` depends on the previous interruption state of the fiber running the effect, which defaults to interruptible.

Reply:
I see. Thank you for the explanation. Yeah I've misread use implementation there.
Ill update the implementations to not restore reserve than to be consistent with this.

In this case it would probably make sense keeping the semaphores of the *parN methods on the acquire, no?

Would it make sense to include an alternative constructor to make that would use acquire for the acquisition instead of reserve. Something like makeInterruptible?

Reply:
The `parN` methods are tricky because when you don't know which step should be limited: `reserve` or `acquire`. I'm not sure what's the best thing to do here. Maybe best to open a dedicated ticket for these too.

Regarding `makeInterruptible`, that's actually expressible using `ZManaged.reserve`: 
```scala
ZManaged.reserve(Reservation(
  acquire = acquireSomethingInterruptibly,
  release = releaseIt
))
```

Reply:
True. I think this is a bit clunky though, personally

Reply:
I agree ;-) But as it is a relatively rare usage of `ZManaged`, I think that's ok.